src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c

```diff
@@ -0,0 +1,328 @@
+    return ctx;
+}
+
+int32_t AndroidCryptoNative_CipherSetTagLength(CipherCtx* ctx, int32_t tagLength)
+{
+    if (!ctx)
+        return FAIL;
+    
+    if(tagLength > TAG_MAX_LENGTH)
+        return FAIL;
+
+    ctx->tagLength = tagLength;
+    return SUCCESS;
+}
+
+static int32_t ReinitializeCipher(CipherCtx* ctx)
+{
+    JNIEnv* env = GetJNIEnv();
+    
+    int32_t keyLength = GetAlgorithmWidth(ctx->type);
+
+    // int ivSize = cipher.getBlockSize();
+    // SecretKeySpec keySpec = new SecretKeySpec(key.getEncoded(), "AES");
+    // IvParameterSpec ivSpec = new IvParameterSpec(IV); or GCMParameterSpec for GCM/CCM
+    // cipher.init(encMode, keySpec, ivSpec);
+
+    jobject algName = GetAlgorithmName(env, ctx->type);
+    if (!algName)
+        return FAIL;
+
+    if (!ctx->ivLength)
+        ctx->ivLength = (*env)->CallIntMethod(env, ctx->cipher, g_getBlockSizeMethod);
+
+    jbyteArray keyBytes = (*env)->NewByteArray(env, keyLength / 8); // bits to bytes, e.g. 256 -> 32
+    (*env)->SetByteArrayRegion(env, keyBytes, 0, keyLength / 8, (jbyte*)ctx->key);
+    jbyteArray ivBytes = (*env)->NewByteArray(env, ctx->ivLength);
+    (*env)->SetByteArrayRegion(env, ivBytes, 0, ctx->ivLength, (jbyte*)ctx->iv);
+
+    jobject sksObj = (*env)->NewObject(env, g_sksClass, g_sksCtor, keyBytes, algName);
+    jobject ivPsObj;
+
+    if (HasTag(ctx->type))
+        ivPsObj = (*env)->NewObject(env, g_GCMParameterSpecClass, g_GCMParameterSpecCtor, ctx->tagLength * 8, ivBytes);
+    else
+        ivPsObj = (*env)->NewObject(env, g_ivPsClass, g_ivPsCtor, ivBytes);
+
+    (*env)->CallVoidMethod(env, ctx->cipher, g_cipherInitMethod, ctx->encMode, sksObj, ivPsObj);
+    (*env)->DeleteLocalRef(env, algName);
+    (*env)->DeleteLocalRef(env, sksObj);
+    (*env)->DeleteLocalRef(env, ivPsObj);
+    (*env)->DeleteLocalRef(env, keyBytes);
+    (*env)->DeleteLocalRef(env, ivBytes);
+
+    return CheckJNIExceptions(env) ? FAIL : SUCCESS;
+}
+
+int32_t AndroidCryptoNative_CipherSetKeyAndIV(CipherCtx* ctx, uint8_t* key, uint8_t* iv, int32_t enc)
+{
+    if (!ctx)
+        return FAIL;
+
+    int32_t keyLength = GetAlgorithmWidth(ctx->type);
+    
+    // input:  0 for Decrypt, 1 for Encrypt, -1 leave untouched
+    // Cipher: 2 for Decrypt, 1 for Encrypt, N/A
+    if (enc != -1)
+    {
+        assert(enc == 0 || enc == 1);
+        ctx->encMode = enc == 0 ? CIPHER_DECRYPT_MODE : CIPHER_ENCRYPT_MODE;
+    }
+
+    // CryptoNative_CipherSetKeyAndIV can be called separately for key and iv
+    // so we need to wait for both and do Init after.
+    if (key)
+        SaveTo(key, &ctx->key, (size_t)keyLength, /* overwrite */ true);
+    if (iv)
+        SaveTo(iv, &ctx->iv, (size_t)ctx->ivLength, /* overwrite */ true);
+
+    if (!ctx->key || !ctx->iv)
+        return SUCCESS;
+
+    return ReinitializeCipher(ctx);
+}
+
+CipherCtx* AndroidCryptoNative_CipherCreate2(CipherInfo* type, uint8_t* key, int32_t keyLength, int32_t effectiveKeyLength, uint8_t* iv, int32_t enc)
```

Comment:
"CipherCreate2", already? :smile:.  If our OpenSSL shim has CIpherCreate2 it's because we created it during complex SxS upgrade time and didn't want to change CipherCreate.

Reply:
I’ll remove the 2 suffix.